src/comparsion.py

```diff
@@ -0,0 +1,143 @@
+    'comma-operator': 'comma_atoms',
+    'implicit-predicate': 'implicit_predicate',
+    'post-increment': 'postIncr',
+    'assignment-as-value': 'assignment_as_value',
+    'repurposed-variable': 'repurposed_variable',
+    'omitted-curly-braces' : 'omitted_curly_braces',
+    'macro-operator-precedence' : 'macro_operator_precedence',
+    'literal-encoding' : 'literal_encoding'
+}
+
+coccinelle_mapping = {
+    'post-increment': 'postIncr',
+    'assignment-as-value': 'assignment_as_value',
+    'conditional' : 'conditional',
+    'pre-increment': 'preIncr',
+    'comma-operator': 'comma_atoms',
+}
+
+def compare_and_output(input_directory, output_directory, codeql_system_path, clojure_system_path, coccinelle_system_path):
+    codeQL_directory = Path(input_directory, "codeQL_data")
+    clojure_directory = Path(input_directory, "clojure_data/atomsClojure.csv")
+    coccinelle_directory = Path(input_directory, "coccinelle_data")
+    dataframes = {}
+
+    # Load CodeQL .sarif files and store them in dataframes dictionary
+
+    for file_path in Path(codeQL_directory).glob('*.sarif'):
+        file_content = file_path.read_text()
+        filename = Path(file_path).stem
+        dataframe_key = filename
+        json_value = json.loads(file_content)
+        p = pd.DataFrame(columns = ["Type", "File", "Line", "CQL", "CQL_Column", "CQL_CodeQL"])
+        rows = []
+        # These are the index paths of the data we are looking for. Each is located inside nested dictionaries.
+        for i in range(len(json_value["runs"][0]["results"])):
+            s = pd.Series({
+                "Type": filename[:-6],
+                "File": json_value["runs"][0]["results"][i]["locations"][0]["physicalLocation"]["artifactLocation"]["uri"], 
+                "Line": json_value["runs"][0]["results"][i]["locations"][0]["physicalLocation"]["region"]["startLine"],
+                "CQL": '+',
+                "CQL_Column": json_value["runs"][0]["results"][i]["locations"][0]["physicalLocation"]["region"].get("startColumn", None),
+                "CQL_Code": json_value["runs"][0]["results"][i]["locations"][0]["physicalLocation"]["contextRegion"]["snippet"]["text"].splitlines()[2:-2],
+            })
+            rows.append(s)
+        p = pd.concat(rows, axis=1).T
+        
+        # These lines convert these columns to string format
+        p["CQL_Code"] = p["CQL_Code"].apply(lambda x: '\n'.join(x) if isinstance(x, list) else x)
+        p["File"] = p["File"].apply(lambda x: '\n'.join(x) if isinstance(x, list) else x)
+        p["File"] = p["File"].str.replace(codeql_system_path, '', 1)
+        # p["File"] = p["File"].str.split('/').str[-1]
+        dataframes[dataframe_key] = p
+            
+    # Load Clojure dataset and print the atoms
+
+    clojure_master = pd.read_csv(clojure_directory)
+    # clojure_master['file'] = clojure_master['file'].str.split('/').str[-1]
+    for i in range(clojure_master["file"].__len__()):   
```

Comment:
This technically works, but it's not very pythonic. Avoid using `__len__`, use `len(list)` instead. A much better way here would be to use `regex`: 
```clojure_master['file'] = clojure_master['file'].apply(
        lambda x: re.sub(f'^{re.escape(clojure_system_path)}', '', x, count=1)
    )
```
Tested and it seems to work